File: tests/fixtures/vuln/nosql_app.java
// tests/fixtures/vuln/nosql_app.java - true-positive fixture for
// modules/sast/rules/nosql.rules (tests/suites/sast.sh), the MongoDB Java
// driver shapes.  One snippet per rule id this file covers, kept well apart so
// no rule's context window leaks into a neighbouring snippet.

import com.mongodb.client.MongoCollection;
import org.bson.Document;

class NoSqlApp {

  Document findByOwner(MongoCollection<Document> coll, String owner) {
    Document filter = Document.parse("{ \"owner\": \"" + owner + "\" }");
    return coll.find(filter).first();
  }

  Document findByName(MongoCollection<Document> coll, String name) {
    Document filter = new Document("$where", "this.name == '" + name + "'");
    return coll.find(filter).first();
  }
}

File: tests/fixtures/vuln/nosql_app.js
// tests/fixtures/vuln/nosql_app.js - true-positive fixture for
// modules/sast/rules/nosql.rules (tests/suites/sast.sh), the MongoDB Node
// driver / Mongoose shapes.  Mirrors tests/fixtures/vuln/app.js's shape: one
// snippet per rule id this file covers, kept well apart so no rule's context
// window leaks into a neighbouring snippet.
//
// Nothing here may reach a rule owned by another pack: this file is walked by
// every shipped SAST pack, not only by nosql.rules.

const { MongoClient } = require('mongodb');

function searchByName(coll, req) {
  const filter = { $where: "this.name == '" + req.query.name + "'" };
  return coll.find(filter);
}

function activeAccounts(coll) {
  return coll.aggregate([{ $match: { $expr: { $function: { body: 'function (s) { return s; }', args: ['$status'], lang: 'js' } } } }]);
}

function login(coll, req) {
  return coll.findOne({ username: req.body.username, password: req.body.password });
}

function filterFromRawQuery(coll, raw) {
  const filter = JSON.parse('{"owner": "' + raw + '"}');
  return coll.find(filter);
}
